torch/csrc/jit/script/inline_loop_condition.cpp

```diff
@@ -0,0 +1,80 @@
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/csrc/jit/ir.h>
+#include <torch/csrc/jit/script/inline_loop_condition.h>
+
+namespace torch {
+namespace jit {
+namespace script {
+
+namespace {
+
+void moveBlockBeforeNode(Node* before_node, Block* block) {
+  for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+    auto block_node = *it++;
+    block_node->moveBefore(before_node);
+  }
+}
+
+} // namespace
+
+// The loop node is initially emitted as:
+// Loop(max_trip_count)
+//    block0(loop_counter) {
+//      <body>
+//    }
+//    block1 {
+//      <loop condition>
+//      -> (condition)
+//    }
+// Here, we inline the loop condition and convert the loop to the form:
+// Loop(max_trip_count, start_condition)
+//    block0(loop_counter, loop_carried_block*) {
+//      <body>
+//       BlockExit(continue_condition, loop_carried_block*)
+//    }
+void inlineLoopCondition(Node* n) {
+  Block* body_block = n->blocks().at(0);
+
+  auto pre_header = n->blocks().at(1);
+  auto header_block = n->addBlock();
```

Comment:
This snippet of code exists in several places and what it does is just copies nodes from `pre_header` before the node `n`. Please factor it into a separate function.

Also, the `header_block` probably should be called `temp_block` or something as it's just a temporary place to clone the nodes to (with that said it might be even inserted to the graph itself, not to the node `n`).